server/SchemaLoader.js

```diff
@@ -0,0 +1,186 @@
+            let content = JSON.stringify(schema);
+            let schemaPath = path.join(this.getSchemasPath(), this.getNameFromSchema(schema));
+
+            // Create folders recursively, if they don't already exist.
+            mkdirp(path.dirname(schemaPath), function(mkdirError) {
+                if (mkdirError) {
+                    console.log(mkdirError);
+                    return;
+                }
+
+                // Save schemas to disk. Always update, schemas might've been updated.
+                fs.writeFile(schemaPath, content, function(writeError) {
+                    if (writeError) {
+                        console.log(writeError);
+                        return;
+                    }
+                });
+            });
+        }, this);
+    }
+
+    getNameFromSchema(schema) {
+        // prettier-ignore
+        return [
+            schema.self.vendor,
+            schema.self.name,
+            schema.self.format,
+            schema.self.version
+        ].join("/");
+    }
+
+    getSchemasPath() {
+        let schemaDir = remote.getGlobal("options").schemaDir;
+        // in production, remove app.asar from the path
+        // cannot use process.resourcesPath in development,
+        // as that will point to electron/dist in node_modules
+        let resourcesPath = remote.app.getAppPath().replace("app.asar", "");
+        return !!schemaDir && path.join(resourcesPath, schemaDir);
+    }
+
+    loadLocalSchemas() {
+        let schemasPath = this.getSchemasPath();
+        let schemas = this.readSchema(schemasPath);
+        logger.info("Schemas loaded from disk");
+        this.emit("schemas-loaded", schemas);
+    }
+
+    readSchema(file) {
+        let stats;
+        try {
+            stats = fs.lstatSync(file);
+        } catch (err) {
+            // catch in case the file or directory could not be resolved,
+            // when e.g. somebody deleted the schemas folder
+            console.log(err);
+            return {};
+        }
+
+        if (stats.isFile() && (path.extname(file) === "" || path.extname(file) === ".json")) {
+            // only accept .json or no extension
+            try {
+                let schema = jsonfile.readFileSync(file);
+                let schemaName = this.getNameFromSchema(schema);
+                return {
+                    [schemaName]: new ValidationSchema(schemaName, schema)
+                };
+            } catch (err) {
+                // catch non-valid JSON schemas
+                console.log(file + ": " + err);
+                return {};
+            }
+        } else if (stats.isDirectory(file)) {
+            let files = fs.readdirSync(file);
+
+            let schemas = {};
+            files.forEach(function(f) {
+                schemas = Object.assign(schemas, this.readSchema(path.join(file, f)));
+            }, this);
+            return schemas;
+        }
+    }
```

Comment:
I assume this method should end in `return {}`, to cover the (implicit) `else` case :)

Reply:
Yikes! Attempt no. 2 ;)

Reply:
Beautiful ;)